shell: Improve handling of log messages

nordic-krch opened the pull request:
If burst of log messages was passed to the shell log
backend, it was likely that messages were lost because
shell had no means to control arrivals of log messages.

Added log message enqueueing timeout to the shell instance
to allow blocking logger thread if short-term arrival rate
exceeded shell capabilities.

Added kconfig option for setting log message queue size
and timeout in RTT and UART instances. Added section in
shell documentation which explains interaction between
the logger and shell instance acting as a logger backend.

Fixes #12040.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
shell: Improve handling of log messages

If burst of log messages was passed to the shell log
backend, it was likely that messages were lost because
shell had no means to control arrivals of log messages.

Added log message enqueueing timeout to the shell instance
to allow blocking logger thread if short-term arrival rate
exceeded shell capabilities.

Added kconfig option for setting log message queue size
and timeout in RTT and UART instances. Added section in
shell documentation which explains interaction between
the logger and shell instance acting as a logger backend.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

jarz-nordic commented:
@nordic-krch : I think you need to update `SHELL_DEFINE` also for dummy backend

jhedberg reviewed (approved):
Works ok for me. I think you're missing "Fixes #12040" from the commit message/description however.

Unrelated to this PR: Any chance we could increase the default log buffer size to 4k? That's the minimum that works for me (together with this PR) so that no logs are missed if I enabled just a single sub-module debug logs (mesh settings in my case). Perhaps it could be made conditional to the chosen SoC or maybe features (like Bluetooth enabled). I'm just worried that we're going to get lots of issue reports from our users who are confused by missing logs (which prior to the new logger were delivered to the console reliably)

dbkinder commented on doc/subsystems/shell/shell.rst:
```diff
@@ -435,4 +435,21 @@ the shell will only print the subcommands registered for this command:
+***************************
+
+Shell instance can act as the :ref:`logger` backend. Shell ensures that log
+messages are correctly mulitplexed with shell output. Log messages from logger
```
```suggestion
 messages are correctly multiplexed with shell output. Log messages from logger
```

dbkinder commented on doc/subsystems/shell/shell.rst:
```diff
@@ -435,4 +435,21 @@ the shell will only print the subcommands registered for this command:
+thread are enqueued and processed in the shell thread. Logger thread will block
+for configurable amount of time if queue is full, blocking logger thread context
+for that time. Oldest log message is removed from the queue after timeout and
+new message is enqueued. `shell stats show` command can be used to retrieve
```
```suggestion
 new message is enqueued. Use the ``shell stats show`` command to retrieve the
```

dbkinder commented on doc/subsystems/shell/shell.rst:
```diff
@@ -435,4 +435,21 @@ the shell will only print the subcommands registered for this command:
+for configurable amount of time if queue is full, blocking logger thread context
+for that time. Oldest log message is removed from the queue after timeout and
+new message is enqueued. `shell stats show` command can be used to retrieve
+number of log message dropped by the shell instance. Log queue size and timeout
```
```suggestion
 number of log messages dropped by the shell instance. Log queue size and timeout
```

dbkinder commented on doc/subsystems/shell/shell.rst:
```diff
@@ -435,4 +435,21 @@ the shell will only print the subcommands registered for this command:
 
+.. warning::
+	Enqueueing timeout must be carefully set if multiple logger backends are
+	used in the system. If timeout is set high to the shell instance which
```
If I understand this right, how about:
```
.. warning::
   Enqueuing timeout must be set carefully when multiple
   backends are used in the system. The shell instance could
   have a slow transport or could block, for example, by a
   UART with hardware flow control. If timeout is set too high,
   the logger thread could be blocked and impact other logger
   backends.
```

pizi-nordic commented on doc/subsystems/shell/shell.rst:
```diff
@@ -435,4 +435,21 @@ the shell will only print the subcommands registered for this command:
+
+Shell instance can act as the :ref:`logger` backend. Shell ensures that log
+messages are correctly multiplexed with shell output. Log messages from logger
+thread are enqueued and processed in the shell thread. Logger thread will block
```
What if logs are processed from user thread or by direct calls to LOG_PROCESS()?

nordic-krch replied:
this context still can be called logger thread context. Do you consider case when LOG_PROCESS() is called from context that cannot be blocked? If so then I can add assert to that call to ensure that message is passed from thread context.

pizi-nordic reviewed (commented):
I think that the proposed implementation might not behave as expected. When incoming message flow will always make the thread blocking, we will expire one message, add to the queue next and start the timeout again significantly reducing speed of the whole logger. IMHO we should expire all the message which are stored in the queue for longer than X.

pizi-nordic commented:
@nordic-krch: Something like this:
```
void add_message_to_queue(message) {
   message->timestamp = k_cycle_get_32() /* TODO: Check if we could reuse logger timestamping */
  fifo_put(message);
}

void flush_queue() {
   now = k_cycle_get_32()
   while(1) {
        message = fifo_peek()
        if (now - message->timestamp < threshold)
           break;

      msg_free(fifo_get());
   }
}
```

nordic-krch committed:
shell: Add dropping of expired messages

Log message get timestamp when being added to shell log message queue.
When adding to log message queue timeouts then all messages added
before timeout are dropped.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
@pizi-nordic take a look. I've added that mechanism as separate commit https://github.com/zephyrproject-rtos/zephyr/pull/12125/commits/ee43e6c445422f6d17d84927b47d7ec336a7be68

pfalcon commented on doc/subsystems/shell/shell.rst:
```diff
@@ -435,4 +435,21 @@ the shell will only print the subcommands registered for this command:
+thread are enqueued and processed in the shell thread. Logger thread will block
+for configurable amount of time if queue is full, blocking logger thread context
+for that time. Oldest log message is removed from the queue after timeout and
+new message is enqueued. Use the ``shell stats show`` command to retrieve
```
I'd like to add that if user needs to check lost messages on their own, that's not ok. The fact that messages are being dropped should be actively shown to the user (as in #12062).

pfalcon commented on subsys/shell/Kconfig.template.shell_log_queue_size:
```diff
@@ -0,0 +1,10 @@
+	int "Log message queue size"
+	default $(default-size)
+	help
+	  Logger threads enqueues log messages which are then processed by shell
```
Log queue size... Again, in which units? For example, log strdup buffer size is measured in units of well, units of log_strdup invocation.

pfalcon commented on subsys/shell/Kconfig.template.shell_log_queue_size:
```diff
@@ -0,0 +1,10 @@
+	  Logger threads enqueues log messages which are then processed by shell
+	  thread. Too small queue may lead to logger thread being blocked
+	  (see $(module)_LOG_MESSAGE_QUEUE_TIMEOUT). Too big queue on relatively
+	  slow shell transport may lead to situation where logs are dropped
```
> Too big queue on relatively
	  slow shell transport may lead to situation where logs are dropped
	  because all log messages are enqueued.

I'm not sure that "too big a queue" would lead to a case of logs being dropped. Likewise, I don't think that logs would be dropped "because all log messages are enqueued.", but because the queue is full. (Then again, the bigger the queue, the lower probability that it becomes full.)

nordic-krch replied:
messages are allocated from the pool in the logger subsystem. If queue in one of the backends is long all allocated messages may stuck in that queue and pool may become empty. In that case logs will be dropped because logger subsystem cannot allocate new messages. If we intent to increase queue size we must take care pool size as well.

pfalcon commented on subsys/shell/Kconfig.template.shell_log_queue_timeout:
```diff
@@ -0,0 +1,10 @@
+	help
+	  If queue with pending log messages is full, oldest log message is
+	  dropped if queue is still full after requested time (-1 is forever).
+	  Logger thread is blocked for that period thus long timeout impacts
```
Should be comma before "thus".

pfalcon commented on include/shell/shell_log_backend.h:
```diff
@@ -40,6 +40,12 @@ struct shell_log_backend {
 
+/** @brief Shell log backend message structure. */
+struct shell_log_backend_msg {
+	struct log_msg * msg;
```
Extra space after "*".

nordic-krch commented:
@carlescufi i think this one also can be merged. I've applied @dbkinder comments.
